src/embed/connection.rs

```diff
@@ -0,0 +1,256 @@
+            req.response
+                .send(Err(format!("Unsupported rpc name {}", req.rpc)))
+                .await
+        }
+    };
+    UnorderedResult::None()
+}
+
+pub async fn process(store: dag::Store, rx: Receiver<Request>) {
+    let txns = RwLock::new(HashMap::new());
+    let mut futures = FuturesUnordered::new();
+    let mut recv = true;
+
+    futures.push(connection_future(&rx, &store, &txns, None));
+    while let Some(value) = futures.next().await {
+        if recv {
+            futures.push(connection_future(&rx, &store, &txns, None));
+        }
+        match value {
+            UnorderedResult::Request(value) => match value {
+                Err(why) => warn!("Dispatch loop recv failed: {}", why),
+                Ok(req) => {
+                    futures.push(connection_future(&rx, &store, &txns, Some(req)));
+                }
+            },
+            UnorderedResult::Stop() => recv = false,
+            UnorderedResult::None() => {}
+        }
+    }
+}
+
+async fn execute<T, F>(func: F, txns: &TxnMap<'_>, req: Request)
+where
+    T: DeJson + TransactionRequest,
+    F: for<'r, 's> AsyncFn2<&'r RwLock<db::Write<'s>>, T, Output = Result<String, String>>,
+{
+    let request: T = match deserialize(&req.data) {
+        Ok(v) => v,
+        Err(e) => return req.response.send(Err(e)).await,
+    };
+
+    let txn_id = request.transaction_id();
+    let txns = txns.read().await;
+    let txn = match txns.get(&txn_id) {
+        Some(v) => v,
+        None => {
+            return req
+                .response
+                .send(Err(format!("No transaction {}", txn_id)))
+                .await
+        }
+    };
+
+    req.response.send(func.call(txn, request).await).await;
+}
+
+async fn do_open<'a, 'b>(store: &'a dag::Store, txns: &'b TxnMap<'a>, req: Request) {
+    use OpenTransactionError::*;
+    let dag_write = store.write().await.map_err(DagWriteError).unwrap();
+    let write = db::Write::new_from_head("main", dag_write)
+        .await
+        .map_err(DBWriteError)
+        .unwrap();
+    let txn_id = TRANSACTION_COUNTER.fetch_add(1, Ordering::SeqCst);
+    txns.write().await.insert(txn_id, RwLock::new(write));
+    req.response
+        .send(Ok(SerJson::serialize_json(&OpenTransactionResponse {
+            transaction_id: txn_id,
+        })))
+        .await;
+}
+
+async fn do_commit(txns: &TxnMap<'_>, req: Request) {
+    let txn_id = match deserialize::<CommitTransactionRequest>(&req.data) {
+        Ok(v) => v.transaction_id,
+        Err(e) => return req.response.send(Err(e)).await,
+    };
+    let mut txns = txns.write().await;
+    let txn = match txns.remove(&txn_id) {
+        Some(v) => v,
+        None => {
+            return req
+                .response
+                .send(Err(format!("No such transaction {}", txn_id)))
+                .await;
+        }
+    };
+    let txn = txn.into_inner();
+    let response = txn
+        .commit(
+            "main",
```

Comment:
@phritz note not all of this fleshed out yet. Just need to add the particulars. to the openTransaction struct and maybe store them in the tx until commit time.

Reply:
K. FWIW I think I'll need the lower level stuff before all the tx functionality working completely.